### PR TITLE
fix(core): skip auto-title generation when history has no user message

### PR DESCRIPTION
**Pull request by yuanyuanAli**
## What this PR does

Adds a guard in `tryGenerateSessionTitle` that checks if the full history contains any user message (`role === 'user'`). If not, returns `empty_history` and skips title generation until a real user message exists.

## Why it's needed

When a daemon session is created and a prompt is sent immediately, the auto-title feature generates a title based on initialization context (e.g., "初始化项目上下文") instead of the user's actual first message (e.g., "你好"). This happens because:

1. Session is created with some initialization content
2. User sends a prompt immediately
3. `maybeTriggerAutoTitle()` fires after the first assistant turn
4. `tryGenerateSessionTitle()` sees enough history (init context + prompt response) and generates a title based on the init content
5. Once the title is set, it's locked — subsequent turns never re-generate

This fix ensures the title is only generated when there's at least one user message in history, preventing titles from being based solely on system initialization.

## Reviewer Test Plan

1. Create a new session and immediately send "你好"
2. Wait for the first assistant turn to complete
3. Verify the auto-generated title is based on the conversation, not initialization context
4. Run the new unit test: `cd packages/core && npx vitest run src/services/chatRecordingService.autoTitle.test.ts -t "fix: skips title generation"`

## Tested on

| OS      | Status |
| ------- | ------ |
| macOS   | ✅     |

## Risk & Scope

- **Risk**: Low — only affects auto-title generation timing
- **Scope**: `packages/core/src/services/sessionTitle.ts` + test
- **Breaking changes**: None
- **Out of scope**: Title quality/coverage heuristics (future improvement)

## Linked Issues

N/A — small bug fix identified during daemon session testing.

### CODE REVIEW

**Review by chiga0 — DISMISSED** [review]
## Review Overview

**PR**: #5120 `fix(core): skip auto-title generation when history has no user message`
**Author**: yuanyuanAli | **HEAD**: `2db48b6` | **+68/-0, 2 files, 1 commit**

### Findings Summary
- **Critical/Major**: 0
- **Minor**: 0
- **Nit**: 1

### Analysis

Surgical fix for a race condition where auto-title generates based on initialization context (system prompts, workspace loading) before the user has sent any message. The guard is correctly placed in `tryGenerateSessionTitle` — after the `fullHistory.length < 2` early-return but before `filterToDialog` — and uses the existing `empty_history` reason, which is semantically correct and already handled by callers (the auto-title trigger retries on subsequent assistant turns).

Key observations:

1. **Guard correctness**: `fullHistory.some((msg) => msg.role === 'user')` is a clean O(n) scan on a small array. The `Content` type from `@google/genai` has `role: 'user' | 'model'`, consistent with the existing `filterToDialog` check.

2. **No regressions**: Once a user message exists, `.some()` always returns true for the rest of the session, so this is purely additive — it only blocks generation in the narrow window before any user message arrives.

3. **Integration test**: The new test case correctly simulates the race condition through the `ChatRecordingService` layer: first mock returns `empty_history` → no title set, second mock (after simulated user message) succeeds → title set to `'问候会话'`.

**Nit**: `tryGenerateSessionTitle` is fully mocked in the test, so the actual `hasUserMessage` guard in `sessionTitle.ts` isn't directly exercised. A unit test for the guard logic itself (e.g., constructing a mock `geminiClient` with only model-role history and asserting `empty_history` is returned) would provide more direct coverage of the code change. Not blocking — the integration test validates the end-to-end behavior.

### Final Verdict — APPROVE

Clean, minimal fix that addresses a real race condition. Correct guard placement, no side effects, well-commented, and covered by an integration test.

---
_This review was generated by QoderWork AI_

File: packages/core/src/services/chatRecordingService.autoTitle.test.ts [hunk 1]
```diff
@@ -549,4 +549,64 @@ describe('ChatRecordingService - auto-title trigger', () => {
     expect(chatRecordingService.getCurrentCustomTitle()).toBe('user-chosen');
     expect(chatRecordingService.getCurrentTitleSource()).toBe('manual');
   });
+
+  it('fix: skips title generation when history has no user message (race condition fix)', async () => {
```
Comment (wenshao) [inline packages/core/src/services/chatRecordingService.autoTitle.test.ts]: **[Suggestion]** The test mocks `tryGenerateSessionTitle` entirely (`mockResolvedValueOnce({ ok: false, reason: 'empty_history' })`), so the actual guard code (`fullHistory.some(msg => msg.role === 'user')` in `sessionTitle.ts`) is never executed by any test. This only verifies that `ChatRecordingService` handles a mocked `empty_history` return — which was already the case before this PR, since `maybeTriggerAutoTitle` does `if (!outcome.ok) return;` for all failure reasons.

A unit test calling the real `tryGenerateSessionTitle` with a model-only history would provide actual regression protection for the guard:

```ts
it('returns empty_history when history has no user message', async () => {
  const { config } = makeConfig({
    fastModel: 'qwen-turbo',
    history: [
      { role: 'model', parts: [{ text: 'Init context loaded.' }] },
      { role: 'model', parts: [{ text: 'Ready to help.' }] },
    ],
  });
  const outcome = await tryGenerateSessionTitle(config, new AbortController().signal);
  expect(outcome).toEqual({ ok: false, reason: 'empty_history' });
});
```

Also, the test name uses a `fix:` prefix and `(race condition fix)` suffix, which is inconsistent with the file's convention — the other 17 tests use plain descriptive names (e.g., `'retries on empty_result up to the cap, then stops'`). Consider renaming to `'skips title generation when history has no user message'`.

_— qwen3.7-max via Qwen Code /review_

File: packages/core/src/services/sessionTitle.ts [hunk 2]
```diff
@@ -116,6 +116,16 @@ export async function tryGenerateSessionTitle(
...
 
+    // Guard: skip title generation if no user message exists in history.
+    // This prevents the title from being generated based solely on
+    // initialization context (e.g., workspace loading, system prompts)
+    // before the user has actually said anything. Without this guard, a
+    // session that sends a prompt immediately after creation could get a
+    // title like "初始化项目上下文" instead of one based on the user's
+    // actual message.
+    const hasUserMessage = fullHistory.some((msg) => msg.role === 'user');
+    if (!hasUserMessage) return { ok: false, reason: 'empty_history' };
```
Comment (doudouOUC) [inline packages/core/src/services/sessionTitle.ts:135]: **[Suggestion]** The new `hasUserMessage` guard has no direct unit test. The test added in `chatRecordingService.autoTitle.test.ts` mocks `tryGenerateSessionTitle` entirely via `vi.mock('./sessionTitle.js')`, so the actual guard code (`fullHistory.some((msg) => msg.role === 'user')`) is never exercised. The existing `empty_history` test in `sessionTitle.test.ts` uses `history: []` (length 0), which exits at the earlier `fullHistory.length < 2` check — it never reaches this branch.

Consider adding a unit test in `sessionTitle.test.ts`:

```typescript
it('returns {ok:false, reason:"empty_history"} when history has only model messages', async () => {
  const { config } = makeConfig({
    fastModel: 'qwen-turbo',
    history: [
      { role: 'model', parts: [{ text: 'Workspace loaded.' }] },
      { role: 'model', parts: [{ text: 'Ready.' }] },
    ],
  });
  const outcome = await tryGenerateSessionTitle(config, new AbortController().signal);
  expect(outcome).toEqual({ ok: false, reason: 'empty_history' });
});
```

_— qwen3.7-max via Qwen Code /review_

File: packages/core/src/services/chatRecordingService.autoTitle.test.ts [hunk 3]
```diff
@@ -549,4 +549,64 @@ describe('ChatRecordingService - auto-title trigger', () => {
...
+      message: [{ text: '已加载项目上下文，准备就绪。' }],
+    });
+
+    await flushMicrotasks();
+
+    // Verify: NO title was generated because history had no user message
+    const titleRecord = findCustomTitleRecord();
+    expect(titleRecord).toBeUndefined();
+    expect(chatRecordingService.getCurrentCustomTitle()).toBeUndefined();
+
```
Comment (doudouOUC) [inline packages/core/src/services/chatRecordingService.autoTitle.test.ts:582]: **[Suggestion]** The first-phase assertions only check that `findCustomTitleRecord()` and `getCurrentCustomTitle()` are `undefined`. Without asserting that `tryGenerateSessionTitleMock` was actually called, this test can't distinguish "guard returned `empty_history`" from "trigger never fired." Every other negative-outcome test in this file asserts on mock call count.

```suggestion
    await flushMicrotasks();

    // Verify: the mock was called and returned empty_history
    expect(tryGenerateSessionTitleMock).toHaveBeenCalledOnce();

    // Verify: NO title was generated because history had no user message
    const titleRecord = findCustomTitleRecord();
    expect(titleRecord).toBeUndefined();
    expect(chatRecordingService.getCurrentCustomTitle()).toBeUndefined();
```

_— qwen3.7-max via Qwen Code /review_

**Review by wenshao — DISMISSED** [review]

**Comment by yuanyuanAli** [comment]
Thanks for the reviews! I have addressed all suggestions:

**@wenshao**
- Added a direct unit test in `sessionTitle.test.ts` that calls the real `tryGenerateSessionTitle` with model-only history
- Renamed the test to match the file naming convention

**@doudouOUC**  
- The new guard is now directly tested in `sessionTitle.test.ts` (see above)
- Added `expect(tryGenerateSessionTitleMock).toHaveBeenCalledOnce()` to verify the mock was actually called

All 34 tests pass (16 in sessionTitle + 18 in autoTitle), lint and typecheck clean.

**Comment by wenshao** [comment]
## 🧪 Local runtime verification (real `tryGenerateSessionTitle`, tmux)

I ran a real‑runtime A/B on this branch (`19f12a40f`) driving the **actual** `tryGenerateSessionTitle` (not mocked) across the three history shapes a fresh session can be in, observing the title‑model boundary (`getBaseLlmClient().generateJson`). The init‑context history was reproduced faithfully from production (`getInitialChatHistory`) and cross‑checked with the project's own helpers.

### TL;DR

> **The guard does not fix the bug it targets, and the new test does not exercise the guard.** The change is harmless (no regression), but ineffective for the described scenario. A one‑line correction using the existing `isSystemReminderContent()` helper *does* fix it (verified below).

**Root cause:** production seeds startup/init context as a **single `role: 'user'` `<system-reminder>` message** (`getInitialChatHistory`). The guard's predicate `fullHistory.some(msg => msg.role === 'user')` is therefore **`true`** for an init‑only history, so the early‑return never fires — the title model is still called with only the init text.

### Evidence 1 — Runtime A/B (does the title model get called? on what input?)

| History when auto‑title fires | **PR guard ON** | **guard OFF** (counterfactual) | **Proposed fix** |
|---|---|---|---|
| **S1** `[user(<system-reminder> startup), model(ack)]` — *the PR's described state* | model **CALLED**, `ok:true` ❌ | model **CALLED**, `ok:true` | model **NOT called**, `empty_history` ✅ |
| **S2** `[model, model]` (zero user‑role entries) | `empty_history`, not called | model CALLED | `empty_history`, not called |
| **S3** `[startup, user("你好"), model]` — real user msg | model called, sees `你好` ✅ | (same) | model called ✅ (no regression) |

For **S1**, *guard ON and guard OFF are identical*. The model input captured at the boundary was:

```
Conversation so far:
User: <system-reminder>
This is the Qwen Code. We are setting up the context for our chat.
Today's date is Monday, June 15, 2026.
My operating system is: linux
I'm currently working in the directory: /home/me/project
</system-reminder>
...
```

i.e. the title model is invoked with **only the startup‑context text** — any title it returns is init‑derived (exactly the reported `初始化项目上下文`). **The guard does not change this.** The *only* shape the guard changes is **S2** (a history with zero `role:'user'` entries), which the documented init flow does not produce (the prelude is always `role:'user'`, and a real prompt is `role:'user'` too).

### Evidence 2 — Structural proof using the project's own helpers

Run against the S1 history `[user(<system-reminder>), model(ack)]`:

| Check | Value |
|---|---|
| `getStartupContextLength(history)` | **1** → it's a startup prelude |
| `isSystemReminderContent(initMsg)` | **true** → structurally a reminder, *not* a turn |
| guard predicate `some(m => m.role==='user')` | **true** → **misclassifies the prelude as a user turn** |
| real user turns after the prelude | **0** |

`getStartupContextLength` / `isSystemReminderContent` already exist for exactly this purpose and are used in `core/client.ts`, `acp-integration/session/Session.ts`, and `ui/utils/historyMapping.ts` to skip the prelude and find the first *real* user turn. The guard reinvents that check incorrectly.

### Evidence 3 — The new test does not test the guard (mutation)

The new test in `chatRecordingService.autoTitle.test.ts` does `vi.mock('./sessionTitle.js')` (line 21), so the real guard **never runs** — it only asserts the caller's behavior given a hard‑coded `{ok:false, reason:'empty_history'}`.

Mutation check (delete the guard from `sessionTitle.ts`, then re‑run):

```
guard present  → 18 passed
guard REMOVED  → 18 passed   ← incl. "fix: skips title generation when history has no user message"
```

The test is green with **or** without the guard — it cannot detect a regression in the code it's meant to protect.

### Verified corrected fix (one line)

Swapping the predicate to exclude the structural prelude makes **S1 skip** (`empty_history`, model not called) while **S3 still generates** (no regression):

```ts
import { isSystemReminderContent } from '../utils/environmentContext.js';
// ...
const hasUserMessage = fullHistory.some(
  (msg) => msg.role === 'user' && !isSystemReminderContent(msg),
);
if (!hasUserMessage) return { ok: false, reason: 'empty_history' };
```

This also handles IDE mode (reminder merged into the prompt → `isSystemReminderContent` is `false` → correctly treated as a real turn) and the legacy `[user(env), model("Got it…")]` pair.

### Recommendation

🔶 **Request changes before merge.** (1) Use `isSystemReminderContent` (or `getStartupContextLength`) so the guard actually skips the startup‑only prelude; (2) make the test drive the **real** `tryGenerateSessionTitle` (as `sessionTitle.test.ts` does) with an init‑only history, asserting `empty_history` + that the model is not called — so it has teeth.

---

<details>
<summary>🇨🇳 中文版（点击展开）</summary>

## 🧪 本地真实运行验证（真实调用 `tryGenerateSessionTitle`，tmux）

我在本分支（`19f12a40f`）上做了真实运行 A/B：直接驱动**真实的** `tryGenerateSessionTitle`（未做 mock），覆盖新会话可能出现的三种历史形态，并在标题模型边界（`getBaseLlmClient().generateJson`）观测调用。初始化上下文历史按生产代码（`getInitialChatHistory`）忠实复现，并用项目自身的辅助函数交叉验证。

### 结论速览

> **该守卫并没有修复它要修的 bug，新增测试也没有真正覆盖该守卫。** 改动无害（无回归），但对所述场景无效。用现有的 `isSystemReminderContent()` 辅助函数改一行即可真正修复（下文已验证）。

**根因：** 生产环境把启动/初始化上下文作为**一条 `role: 'user'` 的 `<system-reminder>` 消息**注入（`getInitialChatHistory`）。因此当历史里只有初始化上下文时，守卫的判断 `fullHistory.some(msg => msg.role === 'user')` 仍为 **`true`**，提前返回永远不会触发——标题模型照样被调用，且输入只有初始化文本。

### 证据 1 — 运行 A/B（标题模型是否被调用？输入是什么？）

| 触发自动标题时的历史 | **PR 守卫开启** | **守卫关闭**（反事实） | **建议修复** |
|---|---|---|---|
| **S1** `[user(<system-reminder> 启动), model(ack)]` —— *PR 描述的场景* | 模型**被调用**，`ok:true` ❌ | 模型**被调用**，`ok:true` | 模型**未调用**，`empty_history` ✅ |
| **S2** `[model, model]`（无任何 user 角色条目） | `empty_history`，未调用 | 模型被调用 | `empty_history`，未调用 |
| **S3** `[启动, user("你好"), model]` —— 含真实用户消息 | 模型被调用，看到 `你好` ✅ | （相同） | 模型被调用 ✅（无回归） |

对 **S1**，*守卫开启与关闭结果完全一致*。在边界处抓到的模型输入为：

```
Conversation so far:
User: <system-reminder>
This is the Qwen Code. We are setting up the context for our chat.
Today's date is Monday, June 15, 2026.
My operating system is: linux
I'm currently working in the directory: /home/me/project
</system-reminder>
...
```

即标题模型的输入**只有启动上下文文本**——它返回的任何标题都来自初始化内容（正是所报的 `初始化项目上下文`）。**守卫并未改变这一点。** 守卫唯一能改变的是 **S2**（历史中完全没有 `role:'user'` 条目）这种形态，而所述初始化流程并不会产生它（前导消息恒为 `role:'user'`，真实用户输入也是 `role:'user'`）。

### 证据 2 — 用项目自身辅助函数的结构性证明

对 S1 历史 `[user(<system-reminder>), model(ack)]` 运行：

| 检查 | 值 |
|---|---|
| `getStartupContextLength(history)` | **1** → 它是启动前导 |
| `isSystemReminderContent(initMsg)` | **true** → 结构上是 reminder，*不是*一轮对话 |
| 守卫判断 `some(m => m.role==='user')` | **true** → **把前导误判为用户轮次** |
| 跳过前导后的真实用户轮数 | **0** |

`getStartupContextLength` / `isSystemReminderContent` 本就是为此而设，已在 `core/client.ts`、`acp-integration/session/Session.ts`、`ui/utils/historyMapping.ts` 中用于跳过前导、定位首个*真实*用户轮次。本守卫用错误的方式重新实现了该判断。

### 证据 3 — 新增测试并未测试守卫（变异测试）

`chatRecordingService.autoTitle.test.ts` 里的新测试执行了 `vi.mock('./sessionTitle.js')`（第 21 行），因此真实守卫**根本不会执行**——它只是在“函数返回写死的 `{ok:false, reason:'empty_history'}`”前提下断言调用方行为。

变异检查（从 `sessionTitle.ts` 删除守卫后重跑）：

```
守卫存在  → 18 passed
守卫删除  → 18 passed   ← 含 “fix: skips title generation when history has no user message”
```

有无守卫测试都绿——它无法发现它本应保护的那段代码的回归。

### 已验证的修复（一行）

把判断改为排除结构性前导后，**S1 会跳过**（`empty_history`，模型不调用），而 **S3 仍正常生成**（无回归）：

```ts
import { isSystemReminderContent } from '../utils/environmentContext.js';
// ...
const hasUserMessage = fullHistory.some(
  (msg) => msg.role === 'user' && !isSystemReminderContent(msg),
);
if (!hasUserMessage) return { ok: false, reason: 'empty_history' };
```

这同样能正确处理 IDE 模式（reminder 被并入 prompt → `isSystemReminderContent` 为 `false` → 正确视为真实轮次）以及旧版 `[user(env), model("Got it…")]` 形态。

### 建议

🔶 **合并前请求修改。**（1）改用 `isSystemReminderContent`（或 `getStartupContextLength`），让守卫真正跳过“仅启动前导”的历史；（2）让测试驱动**真实**的 `tryGenerateSessionTitle`（参照 `sessionTitle.test.ts`），用“仅初始化”历史断言返回 `empty_history` 且模型未被调用——这样测试才有效。

</details>

<sub>Method: real `tryGenerateSessionTitle` exercised via the existing `sessionTitle.test.ts` config seam; provider boundary (`generateJson`) observed like a mock‑OpenAI server. Init history reproduced from `getInitialChatHistory` and validated with `getStartupContextLength`/`isSystemReminderContent`. All runs executed in tmux on `19f12a40f`.</sub>

File: packages/core/src/services/sessionTitle.ts (hunk 2 above)
Comment (wenshao) [inline packages/core/src/services/sessionTitle.ts:135]: **[Suggestion]** The `empty_history` return from this guard causes the caller `maybeTriggerAutoTitle` (chatRecordingService.ts) to silently consume one of its 3 attempts (`AUTO_TITLE_ATTEMPT_CAP=3`). Unlike the pre-existing `length < 2` check (which fires at most once on the very first turn), this guard can fire on *every* assistant turn until a user message appears. In daemon/ACP sessions with 3+ initialization assistant turns before any user message (e.g., workspace loading, tool discovery), all 3 attempts are exhausted on deterministic failures and auto-title is permanently blocked for the session — even after the user eventually sends a message.

Consider either:
(a) Not counting `empty_history` outcomes against the attempt budget — decrement `autoTitleAttempts` when `reason === 'empty_history'` in the caller.
(b) Adding a distinct `'no_user_message'` reason that the caller treats as non-counting ("not ready yet" rather than "tried and failed").

_— qwen3.7-max via Qwen Code /review_

**Review by wenshao — COMMENTED** [review]
⚠️ Downgraded from Approve to Comment: CI still running. The guard and tests are correct — one design consideration below.

_— qwen3.7-max via Qwen Code /review_

**Comment by yuanyuanAli** [comment]
Thanks for the thorough analysis — you're right on both counts.

Fixed in `13de91e74`:

1. **Guard corrected** — added `!isSystemReminderContent(msg)` so the startup prelude (`role:'user'` wrapping `<system-reminder>`) is properly excluded. The guard now short-circuits for S1 (init-only history) as intended.

2. **Test fixed with teeth** — replaced the `[model, model]` history (which production never produces) with the real init shape `[user(<system-reminder>), model]`. Added `expect(generateJson).not.toHaveBeenCalled()` to assert the model is never called. Mutation test confirms the test fails when the guard is removed.

34/34 tests pass, lint + typecheck clean.

**Comment by yuanyuanAli** [comment]
Good point in general. With the corrected guard, the init-only case now short-circuits before consuming a model call — it returns `empty_history` without hitting the LLM, so the attempt count still increments but the cost is negligible (just the O(n) history scan). In practice, by the time 3 attempts are exhausted the user has typically sent a real message, so this shouldn't block title generation. Will keep this in mind if we see edge cases with 3+ init turns.

**Comment by wenshao** [comment]
## ✅ Verification report — A/B behavioral proof + edge-case audit + real-process check

I validated this fix by building it from the branch, running the suite, writing an independent edge-case audit, and proving the behavior change with a source-level **A/B** against the pre-fix base — plus a real `qwen serve` daemon + interactive TUI run (over tmux) to confirm the fix's core assumption against the live product. **TL;DR: clean, correct, well-targeted fix — recommend merge.** Only minor doc/notes below.

### How it was validated

| Area | Method | Result |
|---|---|---|
| Unit tests | `vitest` (the 2 changed test files) | **34 pass** — `sessionTitle.test.ts` 16 (incl. the direct guard test), `chatRecordingService.autoTitle.test.ts` 18 |
| Edge-case audit | independent test, 6 cases | **all pass** — real user proceeds; reminder-only skips; **mixed (real text + appended reminder) counts as real**; multi-reminder+user proceeds; prelude+user proceeds; reminder-only-user excluded |
| **A/B proof** | run the guard tests against the pre-fix base | **base reproduces the bug, PR fixes it** (below) |
| Assumption check | read `getInitialChatHistory` + dump a real daemon/TUI session's in-memory history | confirmed: the startup prelude **is** a `role:'user'` `<system-reminder>` — exactly what the guard excludes |

### The fix, and why it's right

The guard (`sessionTitle.ts:132-135`) is **not** a naive `role === 'user'` check (as the PR description implies) — it is:

```ts
const hasUserMessage = fullHistory.some(
  (msg) => msg.role === 'user' && !isSystemReminderContent(msg),
);
if (!hasUserMessage) return { ok: false, reason: 'empty_history' };
```

This is the crucial detail: the startup prelude (`getInitialChatHistory`) is itself a `role:'user'` message wrapping `<system-reminder>`, so a plain role check would misclassify it as a real user turn and the bug would persist. `isSystemReminderContent` uses `parts.every(...)`, so a message only counts as a reminder when it's *entirely* reminder content — a real message with an appended reminder still counts as a real user turn (my audit confirms this).

### A/B proof (the centerpiece)

Running the guard tests against the **pre-fix base** (`git show 13de91e74^:…/sessionTitle.ts`):

- ❌ base: `[<system-reminder> prelude, assistant]` history → title generation **proceeds** (my audit got `{ok:true}` = it would title from the init prelude; the PR's own test got `model_error` = it called the title model on init). **This is the #5014 bug.**
- ✅ PR: same history → `{ok:false, reason:'empty_history'}`, and the title model is **not called**.
- ✅ control: a real user message, a mixed message, and multi-reminder+user all **proceed** on both base and PR — so the fix does **not** over-suppress.

So the bug demonstrably exists, the fix demonstrably resolves it, and it doesn't regress the normal path.

### Real-process check & methodology note

I ran a real `qwen serve` daemon and a real interactive TUI (tmux) against a mock OpenAI server, created a session, and sent "你好". I confirmed the live in-memory history begins with the `<system-reminder>` prelude (validating the guard's target). The auto-title *title-generation* side-query did **not** fire in my mock harness — auto-title is gated on `isInteractive() || experimentalZedIntegration()` **and** a resolvable fast model, and that trigger didn't engage in the mock setup. That trigger path is unrelated to this PR (which only changes what happens *when* title-gen runs), and the guard's entire behavioral surface is decisively covered by the A/B above with the exact realistic data shape — so end-to-end title-gen-in-a-live-daemon was not separately observed, but the fix is conclusively validated.

### Notes (non-blocking)

- **Doc**: the PR description says the guard "checks `role === 'user'`", but the load-bearing part is the `!isSystemReminderContent` exclusion (the prelude is itself `role:'user'`). Worth updating the description so reviewers see why a plain role check wouldn't work. The code + inline comment are already correct.
- **Test depth**: `chatRecordingService.autoTitle.test.ts` *mocks* `tryGenerateSessionTitle` (it verifies the consumer handles `empty_history`, not the guard). The guard itself is properly tested in `sessionTitle.test.ts` — coverage is adequate, just noting the split.
- **Theoretical edge**: a user message whose content is *entirely* a literal `<system-reminder>…</system-reminder>` block would be excluded (title skipped). Pathological (a user pasting reminder tags) — negligible.

### Recommendation

**Approve / merge.** Small, correct, well-scoped fix with the right mechanism (system-reminder exclusion), no over-suppression, and a direct unit test. Optionally tighten the PR description's mechanism wording.

<sub>Validated at `13de91e74` (delta = that single commit). A/B via `git show 13de91e74^:…/sessionTitle.ts` swap; real run via `node packages/cli serve` + interactive TUI against the repo's mock OpenAI server over tmux. Guard: `sessionTitle.ts:132`; `isSystemReminderContent`: `environmentContext.ts:330`.</sub>

<details>
<summary>🇨🇳 中文版验证报告（点击展开）</summary>

## ✅ 验证报告 —— A/B 行为对比 + 边界审计 + 真实进程检查

我从分支构建了该修复，跑了测试，写了独立的边界用例审计，并用对比 pre-fix base 的源码级 **A/B** 证明了行为变化 —— 另外用真实 `qwen serve` daemon + 交互式 TUI（经 tmux）确认了修复的核心假设在真实产品中成立。**结论：干净、正确、定位准确的修复 —— 建议合并。** 仅有下面几点小注记。

### 验证方式

| 范围 | 方法 | 结果 |
|---|---|---|
| 单测 | `vitest`（2 个改动测试文件） | **34 通过** —— `sessionTitle.test.ts` 16（含直接的 guard 测试）、`chatRecordingService.autoTitle.test.ts` 18 |
| 边界审计 | 独立测试，6 个用例 | **全过** —— 真实用户消息放行；纯 reminder 跳过；**混合（真实文本 + 追加 reminder）算真实**；多 reminder+用户放行；prelude+用户放行；纯-reminder-用户 被排除 |
| **A/B 证明** | 把 guard 测试跑在 pre-fix base 上 | **base 复现 bug，PR 修复**（见下） |
| 假设核对 | 读 `getInitialChatHistory` + dump 真实 daemon/TUI 会话的内存 history | 确认：启动 prelude **就是** 一个 `role:'user'` 的 `<system-reminder>` —— 正是 guard 排除的对象 |

### 修复本身，以及它为什么对

guard（`sessionTitle.ts:132-135`）**不是** PR 描述暗示的朴素 `role === 'user'` 检查，而是：

```ts
const hasUserMessage = fullHistory.some(
  (msg) => msg.role === 'user' && !isSystemReminderContent(msg),
);
if (!hasUserMessage) return { ok: false, reason: 'empty_history' };
```

这是关键：启动 prelude（`getInitialChatHistory`）本身就是一个包着 `<system-reminder>` 的 `role:'user'` 消息，朴素的 role 检查会把它误判为真实用户回合，bug 依旧。`isSystemReminderContent` 用的是 `parts.every(...)`，所以只有当消息**全部**是 reminder 内容时才算 reminder —— 一条带追加 reminder 的真实消息仍算真实用户回合（我的审计确认了这点）。

### A/B 证明（核心）

把 guard 测试跑在 **pre-fix base** 上（`git show 13de91e74^:…/sessionTitle.ts`）：

- ❌ base：`[<system-reminder> prelude, assistant]` 历史 → 标题生成**继续进行**（我的审计得到 `{ok:true}` = 会基于 init prelude 生成标题；PR 自带测试得到 `model_error` = 在 init 上调用了标题模型）。**这就是 #5014 的 bug。**
- ✅ PR：同样的历史 → `{ok:false, reason:'empty_history'}`，且标题模型**不会被调用**。
- ✅ 对照：真实用户消息、混合消息、多 reminder+用户 在 base 与 PR 上都**继续进行** —— 所以修复**不会过度抑制**。

即：bug 确实存在，修复确实解决它，且不回归正常路径。

### 真实进程检查与方法学说明

我用真实 `qwen serve` daemon 和真实交互式 TUI（tmux）对接 mock OpenAI server，建会话并发送"你好"。确认了实时内存 history 以 `<system-reminder>` prelude 开头（验证了 guard 的目标对象）。但 auto-title 的*标题生成* side-query 在我的 mock 环境里**没有触发** —— auto-title 受 `isInteractive() || experimentalZedIntegration()` **且**可解析的 fast model 门控，而该触发在 mock 环境下未启动。该触发路径与本 PR 无关（本 PR 只改变 title-gen *运行时* 的行为），而 guard 的全部行为面已被上面的 A/B 用完全真实的数据形状决定性覆盖 —— 因此"实时 daemon 端到端的标题生成"未单独观测到，但修复已被确凿验证。

### 注记（非阻塞）

- **文档**：PR 描述说 guard "checks `role === 'user'`"，但真正起作用的是 `!isSystemReminderContent` 排除（prelude 本身就是 `role:'user'`）。建议更新描述，让评审者明白为什么朴素 role 检查不行。代码与内联注释本身已正确。
- **测试深度**：`chatRecordingService.autoTitle.test.ts` *mock* 了 `tryGenerateSessionTitle`（验证的是消费方处理 `empty_history`，不是 guard 本身）。guard 本身在 `sessionTitle.test.ts` 里有正确测试 —— 覆盖足够，只是说明这种分工。
- **理论边界**：内容**完全**是字面 `<system-reminder>…</system-reminder>` 块的用户消息会被排除（跳过标题）。极端情况（用户粘贴 reminder 标签）—— 可忽略。

### 建议

**通过 / 合并。** 小而正确、范围清晰的修复，机制对（system-reminder 排除）、不过度抑制、有直接单测。可顺手收紧一下 PR 描述里对机制的措辞。

<sub>基于 `13de91e74` 验证（增量 = 该单个 commit）。A/B 通过 `git show 13de91e74^:…/sessionTitle.ts` 切换；真实运行用 `node packages/cli serve` + 交互式 TUI 对接仓库的 mock OpenAI server，经 tmux。Guard：`sessionTitle.ts:132`；`isSystemReminderContent`：`environmentContext.ts:330`。</sub>

</details>

**Review by wenshao — APPROVED** [review]

**Comment by wenshao** [comment]
@qwen-code /triage

**Review by chiga0 — APPROVED** [review]
## Re-Review at HEAD `13de91e`

HEAD advanced from `2db48b6` → `13de91e` (amended commit). Substantive improvements addressing all reviewer feedback.

### Changes Since Last Review

**1. Guard enhanced with `isSystemReminderContent`** (`sessionTitle.ts`)

The original guard `fullHistory.some(msg => msg.role === 'user')` was too naive — the startup prelude is a `role:'user'` message wrapping `<system-reminder>` content (injected by `getInitialChatHistory`). The new guard:

```typescript
const hasUserMessage = fullHistory.some(
  (msg) => msg.role === 'user' && !isSystemReminderContent(msg),
);
```

Correctly excludes the startup prelude. `isSystemReminderContent` is well-documented: checks that ALL parts are `<system-reminder>…</system-reminder>` wrapped, with specific handling for IDE mode (where the reminder is concatenated into the prompt's text part).

**2. Direct unit test added** (`sessionTitle.test.ts`)

My previous Nit (guard not directly tested) is fully addressed. The new test constructs a history with only the startup prelude + model ack, asserts `empty_history` outcome, AND verifies `generateJson` was NOT called (the guard short-circuits before reaching the LLM).

**3. Integration test hardened** (`chatRecordingService.autoTitle.test.ts`)

Added `expect(tryGenerateSessionTitleMock).toHaveBeenCalledOnce()` — addresses doudouOUC's feedback that the test couldn't distinguish "guard returned empty_history" from "trigger never fired". Also added explicit type assertion for `systemPayload`.

### Previous Finding Status

| Previous Finding | Status | Evidence |
|-----------------|--------|----------|
| My Nit: guard logic not directly unit tested | ✓ Fixed | `sessionTitle.test.ts` lines 78-113 — exercises real `tryGenerateSessionTitle` with mock history |
| wenshao: no direct test for `hasUserMessage` guard | ✓ Fixed | Same test, with realistic startup prelude content |
| doudouOUC: can't distinguish guard from no-trigger | ✓ Fixed | `expect(tryGenerateSessionTitleMock).toHaveBeenCalledOnce()` |
| wenshao: attempt cap consumption | Acknowledged | `autoTitleAttempts++` at line 995 fires before the LLM call (pre-existing design). The guard's `empty_history` consumes 1 of 3 attempts. In practice the startup prelude triggers at most 1 extra attempt. Not blocking. |

### Final Verdict — APPROVE

All reviewer feedback addressed. The `isSystemReminderContent` enhancement makes the guard robust against the actual startup prelude format. Direct unit test + hardened integration test provide solid coverage.

---
_This review was generated by QoderWork AI_
